Avoid permanent cs_main/cs_wallet lock during RescanFromTime

jonasschnelli opened the pull request:
Right now, we are holding `cs_main`/`cs_wallet` during the whole rescan process (which can take a couple of hours).
This was probably only done because of laziness and it is an important show-stopper for #11200 (GUI rescan abort).

jonasschnelli committed:
Avoid pemanent cs_main/cs_wallet lock during wallet rescans

jonasschnelli commented:
Best reviewed with `?w=1` (https://github.com/bitcoin/bitcoin/pull/11281/files?w=1)

practicalswift commented:
@jonasschnelli Which variables are guarded by `cs_main` and `cs_wallet` in these specific cases?

I'll double-check against my lock annotations to make sure they are in sync.

promag commented:
Awesome, will review.

sipa commented on src/wallet/wallet.cpp:
```diff
@@ -1529,14 +1529,15 @@ void CWalletTx::GetAmounts(std::list<COutputEntry>& listReceived,
-    LogPrintf("%s: Rescanning last %i blocks\n", __func__, startBlock ? chainActive.Height() - startBlock->nHeight + 1 : 0);
+    CBlockIndex* startBlock = nullptr;
+    {
+        LOCK2(cs_main, cs_wallet);
```
What needs cs_wallet here?

sipa commented on src/wallet/wallet.cpp:
```diff
@@ -1564,13 +1565,18 @@ CBlockIndex* CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, bool f
+        double dProgressStart = 0;
+        double dProgressTip = 0;
+        {
+            LOCK2(cs_main, cs_wallet);
```
What needs cs_wallet here?

sipa commented on src/wallet/wallet.cpp:
```diff
@@ -1582,13 +1588,17 @@ CBlockIndex* CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, bool f
 
             CBlock block;
             if (ReadBlockFromDisk(block, pindex, Params().GetConsensus())) {
+                LOCK2(cs_main, cs_wallet);
```
What needs cs_main here?

promag replied:
chainActive.

sipa commented on src/wallet/wallet.cpp:
```diff
@@ -1582,13 +1588,17 @@ CBlockIndex* CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, bool f
             }
-            pindex = chainActive.Next(pindex);
+            {
+                LOCK2(cs_main, cs_wallet);
```
What needs cs_wallet here?

jonasschnelli commented:
Updated and reduced locks after @sipa's mentioning.

achow101 commented:
utACK 1d367effa8e7a19303ae46a6b871084cbaaf5a0a

promag commented on src/wallet/wallet.cpp:
```diff
@@ -1564,13 +1565,18 @@ CBlockIndex* CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, bool f
         ShowProgress(_("Rescanning..."), 0); // show rescan progress in GUI as dialog or on splashscreen, if -rescan on startup
-        double dProgressStart = GuessVerificationProgress(chainParams.TxData(), pindex);
-        double dProgressTip = GuessVerificationProgress(chainParams.TxData(), chainActive.Tip());
+        double dProgressStart = 0;
```
Only `chainActive.Tip()` needs the lock? So keep old code and:
```cpp
CBlockIndex* tip = nullptr;
{
    LOCK(cs_main);
    tip = chainActive.Tip();
}

double dProgressStart = GuessVerificationProgress(chainParams.TxData(), pindex);
double dProgressTip = GuessVerificationProgress(chainParams.TxData(), tip);
```

promag replied:
BTW, this patch allows the tip to be update so `dProgressTip` should be inside the loop?

theuni commented on src/wallet/wallet.cpp:
```diff
@@ -1582,13 +1588,17 @@ CBlockIndex* CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, bool f
-            pindex = chainActive.Next(pindex);
+            {
+                LOCK(cs_main);
+                pindex = chainActive.Next(pindex);
```
What if there's a reorg between iterations?

Does anything weird happen if we (for ex): begin scanning, receive a new best block with our tx from p2p, then re-scan here?

Seems to me that we should take a copy of ChainActive, traverse, compare to progress to chainActive.Tip(), and repeat until they match. Something like (just a quick sketch, probably broken):
```c++
CBlockIndex* pindex = pindexStart;
CChain chain;
while (!fAbortRescan) {
    {
        LOCK(cs_main);
        if (pindex == chainActive.Tip()) {
            break;
        }
        chain.SetTip(chainActive.Tip());
    }
    if (pindex) {
        // start where we left off.
        const CBlockIndex* fork = chain.FindFork(pindex);
        if (fork != pindex) {
            // Need to undo the orphans.
        }
        pindex = chain.Next(fork);
    }
    while(!fAbortRescan && pindex) {
        /*
        do stuff
        ...
        pindex = chain.Next(pindex);
        */
    }
    pindex = chain.Tip();
}
```
That almost entirely avoids locking cs_main, while ensuring that we're never jumping chains.

It could also be done with just the pindex and GetAncestor(), to skip the memory overhead of the second chain copy.

jonasschnelli replied:
I'm not sure if this is really a problem. During rescans, IMO it won't hurt to scan an orphaned block nor does it hurt if you do not scan up to the tip in case you have connected new blocks during the time of re-scanning. This, because those new blocks must also have been scanned by the wallet logic (outside of `ScanForWalletTransactions`).

laanwj replied:
If a reorg happens during rescan, we want to continue rescanning using the current chain, not the outdated one.

> This, because those new blocks must also have been scanned by the wallet logic (outside of ScanForWalletTransactions).

Indeed.

promag replied:
Maybe break the rescan as soon as possible if the above tip is not in the current chain? Then add here:
```cpp
if (!chainActive.Contains(tip)) break;
```

jonasschnelli replied:
In case of a reorg, wouldn't `chainActive.Next(pindex);` return a `nullptr` leading to stop the scan, which should be totally fine. During the reorg, the blocks of the new chain must have been scanned by the wallet via ConnectBlock.

promag replied:
Yes that's true when pindex is not in the chain. But if any later block is not in the chain, why keep rescanning?

jonasschnelli replied:
It should be detected during the next `chainActive.Next(pindex);` which, in case we had a fork between the last `chainActive.Next(pindex);` we would only scan max 1 block... also very unlikely IMO.

promag replied:
Suppose the following:
 - `A B C D E F G H I J` - chain to rescan, tip = J
 - `(A) B C D E F G H I J` - scanning block A
 - `A (B) C D E F G H I J` - scanning block B
 - `A (B) C D E F G H K L M` - scanning block B, reorg after H

Should only rescan to H inclusive? Agree it's an edge case.

BTW, if the tip changes `dProgressTip` could be updated?

TheBlueMatt replied:
Yea, I think there is a real issue here. I (believe) it could be fixed by doing the AddToWalletIfInvolvingMe inside a cs_main (which is going to be at least mostly required since ReadBlockFromDisk requires cs_main) with a chainActive.Contains() check before doing the AddToWalletIfInvolvingMe. That should put us back to the original state of anything we add to the wallet is on the main chain and, thus, will either be Disconnected then Connected in the callbacks, only Connected, or not included in a callback at all.

jonasschnelli replied:
I still don't fully understand the concerns (please help me).

* If there is a reorg, all blocks of the new chain will be scanned by the wallet via ConnectBlock (no need to rescan again, we only need to rescan the stuff that doesn't come in via Connect/Disconnect block).
* If we scan blocks that are no longer in the main-chain, and we could find a transaction in those blocks that is relevant to our wallet, so be it. The wallet does display such transactions as unconfirmed... same as we would have a reorg outside of the rescan process.

TheBlueMatt replied:
Heh, sorry, I didnt fully specify my concern because I thought the above discussion did, but I dont think the above comments are correct. As you point out, I do not believe there is a race where you simply miss transactions. However, there is a race where you may end up marking a transaction as in the wrong block:

A -> B -> ....
    -> C -> ...

Both B and C have the transaction you're interested in.

You go to load block B from a rescan, but it has already been reorg'd and the notificatitons for disconnect(B) and connect(C) have already fired. Thus, you have the transaction marked as in block C, which is correct, but then you finish rescanning and mark the transaction as in block B, which now makes you think the transaction has been conflicted and has negative GetDepthInMainChain.

As I mentioned, this is most simply fixed by adding a "LOCK(cs_main); if (!chainActive.Contains(pindex)) break;" outside the AddToWalletIfInvolvingMe loop.

jonasschnelli replied:
Fixed

promag commented:
Sorry for the bad reference ☝️ ...

jonasschnelli commented:
Fixed @promag's code folding/style issue

jonasschnelli commented:
Reviewers welcome (to make progress with GUI rescan abort #11200), best reviewed with ?w=1

laanwj commented:
utACK 726fe69. There might be a slight overhead to locking for every block, likely only relevant for the first blocks of the chain which have few transactions. In any case this is preferable to the current situation.

promag reviewed (commented):
LGTM, will play a bit.

promag commented:
BTW, see question https://github.com/bitcoin/bitcoin/pull/11281#discussion_r139549593.

promag commented:
This also allows concurrent rescans correct? It will mess up the progress dialog no?

laanwj commented:
> This also allows concurrent rescans correct? It will mess up the progress dialog no?

Concurrent rescans on multiple wallets would be nice, multiple concurrent rescans on one wallet would be dangerous.
Progress dialog issues can be fixed later.

jonasschnelli commented:
Rebased the PR.

* Added checks to make sure only one rescan per time and per wallet can be executed (with appropriate error responses)

* Extended the lock reduction to the new `rescanblockchain` call

jonasschnelli commented:
Best reviewed with [w=1](https://github.com/bitcoin/bitcoin/pull/11281/files?w=1).
Easy way to test multiple parallel rescans by adding a sleep within the inner rescan loop.

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -1651,13 +1652,17 @@ CBlockIndex* CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, CBlock
+        }
         double dProgressStart = GuessVerificationProgress(chainParams.TxData(), pindex);
-        double dProgressTip = GuessVerificationProgress(chainParams.TxData(), chainActive.Tip());
+        double dProgressTip = GuessVerificationProgress(chainParams.TxData(), tip);
```
GuessVerificationProcess needs cs_main (probably best to lock cs_main in GuessVerificationProcess).

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -1669,6 +1674,7 @@ CBlockIndex* CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, CBlock
 
             CBlock block;
             if (ReadBlockFromDisk(block, pindex, Params().GetConsensus())) {
+                LOCK(cs_wallet);
```
(one line up) ReadBlockFromDisk in CBlockIndex* form requires cs_main.

TheBlueMatt commented on src/wallet/rpcdump.cpp:
```diff
@@ -428,6 +432,10 @@ UniValue importpubkey(const JSONRPCRequest& request)
     if (fRescan && fPruneMode)
         throw JSONRPCError(RPC_WALLET_ERROR, "Rescan is disabled in pruned mode");
 
+    if (fRescan && pwallet->IsScanning()) {
```
This is very race-y. You could check check this, have another thread call rescan, and then end up with two parallell rescans.

promag commented:
> > This also allows concurrent rescans correct? It will mess up the progress dialog no?
>
> Concurrent rescans on multiple wallets would be nice, multiple concurrent rescans on one wallet would be dangerous or at least counter-productive (so we might want to track that?).
> Progress dialog issues can be fixed later.

Please see #11826.

ryanofsky commented:
Seems like this PR still has some races described above that need to be fixed.

Also, it is unclear to me what the "important show-stopper for #11200" mentioned in the PR description is. The comments in #11200 seems to indicate that there is some existing deadlock that this fixes and #11200 would worsen without this PR. But there's no explanation of how or when the deadlock happens. It's be good to have some clarification about this. (I also asked about it in https://github.com/bitcoin/bitcoin/pull/11200#issuecomment-340865856).

jonasschnelli commented:
@ryanofsky: #11200 won't currently work because the rescan locks `cs_main` over a longer period. The GUI also tries to lock `cs_main` via graphical updates on the transaction table as well as the balances.... == unresponsiveness (not actually a deadlock) == you won't be able to abort the rescan because the whole GUI is locked.

jonasschnelli commented:
Added two commits.
* Addressed the race issue found by @promag and @TheBlueMatt (d7f3836b74b790887793997562216885840df7a2)
* Update the progress max if the tip has changed (as found by @promag, 9b59847ddca46d30404ca574cba32b6de214844f)

jonasschnelli commented:
Added a commit that fixes the remaining missing locks mentioned by @TheBlueMatt, expect the racy `IsScanning()`. Would it make sense to read and set (protected by a internal lock) the scanning flag right after checking (something like `checkAndEventuallySetScanning()`)? Other ideas?

TheBlueMatt commented:
re: race: I certainly appreciate the RPC exception before importing in case of an existing scan. It might be nice to get some RAII-like lock on the ability to scan, ie like your suggestion but you create a class instance which gives you the right to rescan instead of having to make sure you set some boolean to false if you decide you don't want to rescan.

jonasschnelli committed:
Add RAII wallet rescan reserver

jonasschnelli commented:
Added an RAII object to preserve wallet rescan races.

TheBlueMatt commented:
It appears test_bitcoin.exe hung. Not sure if PR-related or some other issue, but I'm not aware of any issues with such hangs on master atm?

jonasschnelli commented:
@TheBlueMatt: where did it hung? Travis seems to be happy?

jonasschnelli commented:
Rebased and cleaned up commit history.

TheBlueMatt commented:
It may be nice to make a WalletRescanReserver a required argument to ScanForWalletTransactions/RescanFromTime to enforce correctness.

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -1703,6 +1703,7 @@ CBlockIndex* CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, CBlock
             {
                 LOCK(cs_main);
                 pindex = chainActive.Next(pindex);
+                if (pindex && !chainActive.Contains(pindex)) break;
```
This check needs to be inside a LOCK2(cs_main, cs_wallet) in the same block as AddToWalletIfInvolvingMe.

jonasschnelli replied:
Fixed.

jonasschnelli committed:
Make sure WalletRescanReserver has successfully reserved the rescan

jonasschnelli commented:
Followed @TheBlueMatt advice and made the `WalletRescanReserver` an argument of `ScanForWalletTransactions` and `RescanFromTime` to ensure on code level that the wallet rescan is reserved before scanning.

TheBlueMatt commented:
See comment github is hiding at https://github.com/bitcoin/bitcoin/pull/11281#discussion_r156484165. Also, could you squash "Make sure we rescan only blocks in the main chain" into the first commit? Otherwise the first commit is "wrong", which is annoying in review.

jonasschnelli committed:
Reduce cs_main lock in ReadBlockFromDisk, only read GetBlockPos under the lock

jonasschnelli commented:
Added a commit ff54709 to allow to call `ReadBlockFromDisk()` without holding `cs_main` (push the lock down for `CDiskBlockPos::GetBlockPos()`).

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -1684,7 +1697,15 @@ CBlockIndex* CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, CBlock
+            {
+                LOCK(cs_main);
+                pindex = chainActive.Next(pindex);
+                if (pindex && !chainActive.Contains(pindex)) break;
```
This line is a noop - chainActive.Next() is never going to return something non-nullptr that isn't on the best chain.

jonasschnelli replied:
Oops. This is a rebase issue, ... removing this line now.

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -1645,36 +1646,48 @@ int64_t CWallet::RescanFromTime(int64_t startTime, bool update)
             CBlock block;
             if (ReadBlockFromDisk(block, pindex, Params().GetConsensus())) {
+                LOCK2(cs_main, cs_wallet);
+                if (pindex && !chainActive.Contains(pindex)) break;
```
There is a weird case here that I think should be documented - something in the description of ScanForWalletTransactions should mention that, in order to ensure scan is "successful", pindex *must* have been/be on the main chain at some point after the addition of any keys which we want to rescan for.

jonasschnelli replied:
Yes. That case is a bit wired. We should probably return a custom state object instead a single CBlockIndex in the long run.
Mentioned that behaviour now in the `ScanForWalletTransactions` comment.

TheBlueMatt commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3376,30 +3376,39 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
-        }
-    }
+    CBlockIndex *pChainTip = nullptr;
+    {
```
I think you meant to put a LOCK(cs_main) at the top of this scope?

jonasschnelli replied:
Oh. Indeed! Added. Fixed.

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -1684,7 +1697,15 @@ CBlockIndex* CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, CBlock
+                if (pindex && !chainActive.Contains(pindex)) break;
+                if (tip != chainActive.Tip()) {
+                    // in case the tip has changed, update progress max
+                    dProgressTip = GuessVerificationProgress(chainParams.TxData(), tip);
```
err, didnt you mean to set tip to chainActive.Tip() first?

jonasschnelli replied:
Argh. Thanks. Fixed.

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -1644,37 +1645,52 @@ int64_t CWallet::RescanFromTime(int64_t startTime, bool update)
  * If pindexStop is not a nullptr, the scan will stop at the block-index
  * defined by pindexStop
+ *
+ * An additional check if the returned CBlockIndex* is/was on the main chain
```
I dont think that fully captures the requirement here. Its actually rather loose, y ou just need to make sure pindexStop (and pindexStart) are on the main chain *after* to the addition of any new keys you want to detect transactions for. What is returned actually doesnt matter because if pindexStart and pindexStop were on the main chain at some point when all the neccessary keys were in the wallet, any missed blocks during rescan were also caught in the validationinterface callbacks.

jonasschnelli replied:
Update the comment.

TheBlueMatt reviewed (commented):
utACK eba53080666402bba2efe4a430bde2b824edef9d except the one minor missing lock.

sipa commented on src/wallet/rpcdump.cpp:
```diff
@@ -101,56 +101,59 @@ UniValue importprivkey(const JSONRPCRequest& request)
-            pwallet->RescanFromTime(TIMESTAMP_MIN, true /* update */);
         }
     }
+    if (fRescan) {
```
Releasing cs_wallet here means another RPC could occur in between which observes the key being present, but not the relevant transactions. Is that a concern?

TheBlueMatt replied:
I wouldnt think so directly. What may be of larger concern may be that you could get a bogus balance due to having found old transactions but not newer ones. Its probably still fine, but we should note it in the help doc for the rescanning functions. Alternatively we could block other wallet RPC calls until rescan finishes.

jonasschnelli replied:
Possible solutions:
1.) User must take care: report the rescan status in `getwalletinfo()`
2.) Block relevant RPC calls during an active rescan...

I think 1) is more flexible but my introduce some pitfalls...

promag replied:
@sipa that is already possible: `importprivekey rescan=false` followed by `rescanblockchain`.

sipa replied:
@promag If you sent rescan to false you're taking that risk yourself obviously. My point is that if you do request a rescan, there is a reasonable expectation to not see inconsistent results.

promag replied:
Yeah I know, there is also reasonable expectation to have the RPC responsive 😄

jonasschnelli commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3376,30 +3376,40 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
     // We can't rescan beyond non-pruned blocks, stop and throw an error
     if (fPruneMode) {
-        CBlockIndex *block = pindexStop ? pindexStop : chainActive.Tip();
+        CBlockIndex *block = pindexStop ? pindexStop : pChainTip;
```
Right!
Fixed by adding a `cs_main` lock for this block (only occurs for pruned peers).

heri99 commented:
Sory for that

On 16 Dec 2017 5:24 am, "Matt Corallo" <notifications@github.com> wrote:

> *@TheBlueMatt* commented on this pull request.
> ------------------------------
>
> In src/wallet/rpcdump.cpp
> <https://github.com/bitcoin/bitcoin/pull/11281#discussion_r157302011>:
>
> >          }
>      }
> +    if (fRescan) {
>
> I wouldnt think so directly. What may be of larger concern may be that you
> could get a bogus balance due to having found old transactions but not
> newer ones. Its probably still fine, but we should note it in the help doc
> for the rescanning functions. Alternatively we could block other wallet RPC
> calls until rescan finishes.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/11281#discussion_r157302011>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AT0amrJB8ZTspbPKK86FyFNuKI74wYBeks5tAuN3gaJpZM4PQhYL>
> .
>

promag commented on src/wallet/wallet.h:
```diff
@@ -1218,4 +1221,31 @@ bool CWallet::DummySignTx(CMutableTransaction &txNew, const ContainerType &coins
+    bool m_could_reserve;
+    explicit WalletRescanReserver(CWalletRef w): m_wallet(w), m_could_reserve(false) {}
+
+    bool reserve() {
```
Nit, `{` in new line.

promag commented on src/wallet/wallet.h:
```diff
@@ -1218,4 +1221,31 @@ bool CWallet::DummySignTx(CMutableTransaction &txNew, const ContainerType &coins
+public:
+    CWalletRef m_wallet;
+    bool m_could_reserve;
+    explicit WalletRescanReserver(CWalletRef w): m_wallet(w), m_could_reserve(false) {}
```
Nit, missing space before `:`.

promag commented on src/wallet/wallet.h:
```diff
@@ -1218,4 +1221,31 @@ bool CWallet::DummySignTx(CMutableTransaction &txNew, const ContainerType &coins
+class WalletRescanReserver
+{
+public:
+    CWalletRef m_wallet;
```
Make these private?

jonasschnelli replied:
Done.

promag commented on src/wallet/wallet.h:
```diff
@@ -656,6 +656,9 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
     static std::atomic<bool> fFlushScheduled;
     std::atomic<bool> fAbortRescan;
     std::atomic<bool> fScanningWallet;
+    std::mutex mutexScanning;
```
Why the mutex when `fScanningWallet` is `std::atomic`? Could use `std::atomic<bool>::compare_exchange_strong`?

jonasschnelli replied:
Yes. Maybe. But lets keep this for now.

jonasschnelli commented:
Fixed @promag's nits

promag commented:
Squash * Make sure WalletRescanReserver has successfully reserved the rescan*  into * Add RAII wallet rescan reserver*?

TheBlueMatt commented:
Needs rebase. Also I believe still needs (at least) doc fixes to address @sipa's comments at https://github.com/bitcoin/bitcoin/pull/11281#discussion_r157300953

promag reviewed (commented):
Needs rebase.

jonasschnelli committed:
Mention that other RPC calls report keys as "imported" while txns are still missing

jonasschnelli commented:
Rebased and added a commit that adds mentioning of the concern reported by @sipa (https://github.com/bitcoin/bitcoin/pull/11281#discussion_r157300953, other RPC calls may report keys as imported but related transactions are still not there because of the ongoing rescan)

promag commented on src/wallet/wallet.h:
```diff
@@ -1218,4 +1222,37 @@ bool CWallet::DummySignTx(CMutableTransaction &txNew, const ContainerType &coins
+    }
+
+    bool isReserved() const {
+        return (m_could_reserve && m_wallet->fScanningWallet);
```
`return m_could_reserve` is enough? `WalletRescanReserver` is the "owner.

jonasschnelli replied:
I think it's better for future changes to keep the check here... don't hurt but may protect from a future mistake.

ryanofsky replied:
In commit "Make sure WalletRescanReserver has successfully reserved".

Alternative could be to just `return m_could_reserve` and assert `!m_could_reserve || fScanningWallet`, so if there is a future mistake, it can be detected instead of papered over.

Also formatting here is inconsistent, opening brace for function goes on new line.

promag commented on src/wallet/wallet.h:
```diff
@@ -1218,4 +1222,37 @@ bool CWallet::DummySignTx(CMutableTransaction &txNew, const ContainerType &coins
+
+    bool reserve()
+    {
+        std::lock_guard<std::mutex> lock(m_wallet->mutexScanning);
```
Add `assert(!m_could_reserve);`.

jonasschnelli replied:
Added

promag commented on src/wallet/wallet.cpp:
```diff
@@ -1639,37 +1640,53 @@ int64_t CWallet::RescanFromTime(int64_t startTime, bool update)
     {
-        LOCK2(cs_main, cs_wallet);
         fAbortRescan = false;
         fScanningWallet = true;
```
Remove, since reserver already set `fScanningWallet` to true.

promag replied:
Should also remove `fScanningWallet = false` at the end?

jonasschnelli replied:
Ah. I see your point. Removed those sets and mentioned that `fScanningWallet` is controlled by `WalletRescanReserver`.

promag commented on src/wallet/wallet.h:
```diff
@@ -656,6 +657,9 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
     static std::atomic<bool> fFlushScheduled;
     std::atomic<bool> fAbortRescan;
     std::atomic<bool> fScanningWallet;
```
Add a comment saying this is controlled by `WalletRescanReserver`.

jonasschnelli replied:
It's not really controlled by `WalletRescanReserver`. This was added before and `ScanForWalletTransactions` sets it to `true`.

TheBlueMatt commented:
Needs rebase and to address @sipa's note. Personally I think just documenting the case is sufficient.

jonasschnelli commented:
Rebased (non trivial).
Another review would help. Thanks

TheBlueMatt commented on src/wallet/rpcdump.cpp:
```diff
@@ -101,61 +102,63 @@ UniValue importprivkey(const JSONRPCRequest& request)
+        if (!key.IsValid()) throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Private key outside allowed range");
 
-        // whenever a key is imported, we need to scan the whole chain
-        pwallet->UpdateTimeFirstKey(1);
```
This line didn't make it into the new rebase, I believe.

jonasschnelli replied:
Oh. Yes. Thanks Will fix

jonasschnelli replied:
Fixed.

TheBlueMatt replied:
nit: Why'd you move it up in the function instead of keeping its position?

TheBlueMatt commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3430,21 +3440,23 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
-        if (pwallet->IsAbortingRescan()) {
-            throw JSONRPCError(RPC_MISC_ERROR, "Rescan aborted.");
+    CBlockIndex *stopBlock = nullptr;
+    {
```
nit: why scope this? There's no lock to scope, could you not just leave this as-is and only change the chainActive.Tip() call?

jonasschnelli replied:
Your right,.. removed the unnecessary scope.

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -1656,24 +1661,40 @@ CBlockIndex* CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, CBlock
+            LOCK(cs_main);
+            tip = chainActive.Tip();
+            dProgressStart = GuessVerificationProgress(chainParams.TxData(), pindex);
+            dProgressTip = GuessVerificationProgress(chainParams.TxData(), tip);
```
Not changed in this PR, but it may be nice (here or in a follow-up before we start showing rescan as a progress bar in more places) to not use the tip here and instead use pindexStop. It makes sense to always give progress towards tip if there are (potentially) API clients which are calling with a stop a ways back from our tip.

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -1683,7 +1704,15 @@ CBlockIndex* CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, CBlock
+                LOCK(cs_main);
+                pindex = chainActive.Next(pindex);
+                if (tip != chainActive.Tip()) {
+                    tip = chainActive.Tip();
```
Same goes for here - while this check makes the behavior not change, I'm not sure the current behavior is exactly optimal...we scan until pindexStop, *not* chainActive.Tip().

TheBlueMatt commented on src/wallet/rpcdump.cpp:
```diff
@@ -86,7 +86,8 @@ UniValue importprivkey(const JSONRPCRequest& request)
             "3. rescan               (boolean, optional, default=true) Rescan the wallet for transactions\n"
-            "\nNote: This call can take minutes to complete if rescan is true.\n"
+            "\nNote: This call can take minutes to complete if rescan is true, during that time, other rpc calls\n"
+            "may report that the imported key exists but related transactions are still missing.\n"
```
You should probably add a ", leading to temporarily incorrect/bogus balances and unspent outputs until rescan completes" to the warning message.

jonasschnelli replied:
Fixed by adding @TheBlueMatt's proposed sentence-part.

TheBlueMatt reviewed (commented):
Aside from needing to add a similar warning message to rescanblockchain, utACK 0dad810adeb3bbb6a60b722240ee2b90136e8384. There are a bunch of comments here, but most of them can be addressed in a follow-up just fine.

Sjors commented:
Not sure what exactly to test here, but at least I can confirm `bitcoin-cli abortrescan` still works and there's no noticeable difference in QT (I assume that's good).

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1656,24 +1661,40 @@ CBlockIndex* CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, CBlock
+            }
+            if (readRet) {
+                LOCK2(cs_main, cs_wallet);
+                if (pindex && !chainActive.Contains(pindex)) break;
```
In commit "Avoid permanent cs_main"

Because of the break here, the "Returns null if scan was successful" documentation is no longer correct. Would suggest changing so it's still easily possible to distinguish success from failure:

```c++
if (pindex && !chainActive.Contains(pindex)) {
    ret = pIndex;
    break;
}
 
```

Also, the reason for this check is obscure enough (https://github.com/bitcoin/bitcoin/pull/11281#discussion_r155326447) that it I think it deserves some kind of comment. I would suggest something like "Abort scan if current block is no longer active, to prevent marking transactions as coming from the wrong block."

jonasschnelli replied:
Nice catch! Thanks. Fixed.

ryanofsky reviewed (commented):
utACK 7aeeb11a2e6fe2671973ef18f556dbaf32ad5f81. Would be nice to see this merged.

I left various comments, and I think some of them are important enough that they should be followed up on, but this can happen in future PRs.

jonasschnelli commented on src/wallet/wallet.cpp:
```diff
@@ -1614,14 +1614,15 @@ void CWalletTx::GetAmounts(std::list<COutputEntry>& listReceived,
  */
 int64_t CWallet::RescanFromTime(int64_t startTime, bool update)
 {
-    AssertLockHeld(cs_main);
```
I don't think we should enforce *not* locking cs_main at this point.

jonasschnelli commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3436,15 +3446,14 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
     UniValue response(UniValue::VOBJ);
     response.pushKV("start_height", pindexStart->nHeight);
-    response.pushKV("stop_height", stopBlock->nHeight);
+    response.pushKV("stop_height", stopBlock ? stopBlock->nHeight : 0);
```
Yes. Your right. Fixed (removed the extra nullptr check).

jonasschnelli commented on src/wallet/wallet.cpp:
```diff
@@ -1643,6 +1644,10 @@ int64_t CWallet::RescanFromTime(int64_t startTime, bool update)
+ *
+ * Caller needs to make sure pindexStop (and the optional pindexStart) are on
+ * the main chain after to the addition of any new keys you want to detect
+ * transactions for.
```
I keep that comment for now and leave it open for another PR. The comment was directly proposed (word by word) by @TheBlueMatt.

jonasschnelli commented on src/wallet/wallet.cpp:
```diff
@@ -1656,24 +1661,40 @@ CBlockIndex* CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, CBlock
         {
-            if (pindex->nHeight % 100 == 0 && dProgressTip - dProgressStart > 0.0)
+            if (pindex->nHeight % 100 == 0 && dProgressTip - dProgressStart > 0.0) {
+                LOCK(cs_main);
```
Good pont. Fixed.

promag reviewed (commented):
Tested ACK 7f81250, some minor comments though.

Agree with @ryanofsky, for 0.16 it should be merged rather sooner than later.

promag commented:
Forgot to mention, it is still not possible to abort the rescan at initialisation time, I guess the goal is to not change that.

laanwj commented:
> Forgot to mention, it is still not possible to abort the rescan at initialisation time, I guess the goal is to not change that.

I don't think that's a goal here?

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1643,37 +1644,60 @@ int64_t CWallet::RescanFromTime(int64_t startTime, bool update)
+            if (pindex->nHeight % 100 == 0 && dProgressTip - dProgressStart > 0.0) {
+                double gvp = 0;
+                {
+                    LOCK(cs_main);
```
@jonasschnelli, is this lock necessary? It seems like either this should be removed, or else a lock needs to be added below at line 1721 (https://github.com/bitcoin/bitcoin/pull/11281/files#diff-b2bb174788c7409b671c46ccc86034bdR1721) if cs_main is required to call GuessVerificationProgress.

jonasschnelli replied:
I think GVP needs cs_main (@TheBlueMatt?), so, yes, your right, L1721 should also have one... working on a fixup.

ryanofsky replied:
Great, thanks for addressing this.

> I think GVP needs cs_main

I was curious about this. It would be good for GuessVerificationProgress comment to mention that it needs cs_main, since it isn't really obvious one way or the other.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1683,14 +1707,20 @@ CBlockIndex* CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, CBlock
-            pindex = chainActive.Next(pindex);
+            {
+                LOCK(cs_main);
+                pindex = chainActive.Next(pindex);
```
I think there is still a bug here where scan will return success (`nullptr`) if it gets aborted due to a reorg.

Maybe this line:

```c++
pindex = chainActive.Next(pindex);
```

should be changed to something like:

```c++
if (CBlockIndex* new_index = chainActive.Next(pindex)) {
    pindex = new_index;
} else {
    /* aborting, reorg */
    ret = pindex;
    break;
}
```

Another alternative could be to resume instead of aborting on reorgs.

ryanofsky replied:
> should be changed to something like:

I think my suggestion above won't work because it will return an error when pindex points to the the tip. Maybe simplest fix would be to restructure the code a little to just acquire cs_main once after ReadBlockFromDisk instead of acquiring, releasing, then immediately re-acquiring. That way there is would be no chance of reorg between the `chainActive.Contains` check above and this `chainActive.Next` call.

ryanofsky replied:
> I think there is still a bug here where scan will return success

Sorry for the churn, but I no longer think there's a bug here. Opened #12275 to try to clarify the return value.
